### PR TITLE
gen-assembly read from private nightlies

### PR DESCRIPTION
**Pull request by ashwindasr**
This change will allow get_nightlies to auto pick the set of consistent nightlies from private release controller

Tests:
- works with private nightlies [correctly](https://art-jenkins.apps.prod-stable-spoke1-dc-iad2.itup.redhat.com/job/aos-cd-builds/job/build%252Fgen-assembly/197/)
- works with public nightlies [correctly](https://art-jenkins.apps.prod-stable-spoke1-dc-iad2.itup.redhat.com/job/aos-cd-builds/job/build%252Fgen-assembly/198/)

Needs rebase if https://github.com/openshift-eng/art-tools/pull/924 is merged first

### CODE REVIEW

**Comment by ashwindasr** [comment]
`make test` passed on local...

File: doozer/doozerlib/cli/get_nightlies.py [hunk 1]
```diff
@@ -187,12 +194,31 @@ async def find_rc_nightlies(runtime: Runtime, arches: Set[str], allow_pending: b
...
         # retrieve the list of nightlies from the release-controller
-        rc_url: str = f"{rc_api_url(tag_base, _arch)}/tags"
+        rc_url: str = f"{rc_api_url(tag_base, _arch, private_nightly)}/tags"
         logger.info(f"Reading nightlies from {rc_url}")
 
-        async with aiohttp.ClientSession() as session:
-            async with session.get(rc_url) as resp:
-                data = await resp.json()
+        # Get the token
+        oc_token_result = subprocess.run(["oc", "whoami", "-t"], capture_output=True, text=True)
```
Comment (thegreyd) [inline doozer/doozerlib/cli/get_nightlies.py:201]: we can move this to inside the if block

Comment (thegreyd) [reply doozer/doozerlib/cli/get_nightlies.py:201]: also let's use exectools instead of subprocess.run ?

File: doozer/doozerlib/cli/get_nightlies.py [hunk 2]
```diff
@@ -243,6 +269,16 @@ def rc_api_url(tag: str, arch: str) -> str:
...
+        arch = f"-{arch}"
+        if arch == "-amd64":
+            # We do not need to add the arch to URL for amd64
+            return f"{constants.RC_BASE_PRIV_URL.format(arch='')}/api/v1/releasestream/{tag}-priv"
+
+        # For other arches, add it to the URL
+        return f"{constants.RC_BASE_PRIV_URL.format(arch=arch)}/api/v1/releasestream/{tag}{arch_suffix}-priv"
+
+    # For public nightlies
     return f"{constants.RC_BASE_URL.format(arch=arch)}/api/v1/releasestream/{tag}{arch_suffix}"
```
Comment (thegreyd) [inline doozer/doozerlib/cli/get_nightlies.py:282]:
```suggestion
    arch_suffix = go_suffix_for_arch(arch, private_nightly)
    return f"{constants.RC_BASE_URL.format(arch=arch)}/api/v1/releasestream/{tag}{arch_suffix}"
```

File: doozer/doozerlib/cli/get_nightlies.py [hunk 3]
```diff
@@ -243,6 +269,16 @@ def rc_api_url(tag: str, arch: str) -> str:
     """
     arch = go_arch_for_brew_arch(arch)
     arch_suffix = go_suffix_for_arch(arch)
```
Comment (thegreyd) [inline doozer/doozerlib/cli/get_nightlies.py:271]: I noticed that we have private bool in the func which we can use
```suggestion
    arch_suffix = go_suffix_for_arch(arch, private_nightly)
```

File: doozer/doozerlib/cli/get_nightlies.py [hunk 4]
```diff
@@ -243,6 +269,16 @@ def rc_api_url(tag: str, arch: str) -> str:
...
+    if private_nightly:
+        arch = f"-{arch}"
+        if arch == "-amd64":
+            # We do not need to add the arch to URL for amd64
+            return f"{constants.RC_BASE_PRIV_URL.format(arch='')}/api/v1/releasestream/{tag}-priv"
+
+        # For other arches, add it to the URL
+        return f"{constants.RC_BASE_PRIV_URL.format(arch=arch)}/api/v1/releasestream/{tag}{arch_suffix}-priv"
```
Comment (thegreyd) [inline doozer/doozerlib/cli/get_nightlies.py:279]:
```suggestion
        return f"{constants.RC_BASE_PRIV_URL.format(arch=arch)}/api/v1/releasestream/{tag}{arch_suffix}"
```

File: doozer/doozerlib/constants.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 RC_BASE_URL = "https://{arch}.ocp.releases.ci.openshift.org"
+RC_BASE_PRIV_URL = "https://openshift-release{arch}-priv.apps.ci.l2s4.p1.openshiftapps.com"
```
Comment (thegreyd) [inline doozer/doozerlib/constants.py:2]:
```suggestion
RC_BASE_PRIV_URL = "https://{arch}.ocp.internal.releases.ci.openshift.org/"
```

File: doozer/doozerlib/cli/get_nightlies.py [hunk 6]
```diff
@@ -187,12 +194,31 @@ async def find_rc_nightlies(runtime: Runtime, arches: Set[str], allow_pending: b
...
+                    raise ValueError("Token empty, might not be logged in to correct cluster")
+            else:
+                raise ValueError("Error while trying to get the token for reading private nightlies")
+
+            async with aiohttp.ClientSession() as session:
+                headers = {
+                    "Authorization": f"Bearer {token}"
+                }
+                async with session.get(rc_url, headers=headers) as resp:
+                    data = await resp.json()
```
Comment (thegreyd) [inline doozer/doozerlib/cli/get_nightlies.py:217]: since this part is being repeated, we can populate `headers` var in `if private_nightly` and send headers in the common code block.

File: doozer/doozerlib/cli/get_nightlies.py (hunk 3 above)
Comment (ashwindasr) [reply doozer/doozerlib/cli/get_nightlies.py:271]: so the base URL for private nightlies is a bit different. from non-amd64, the arch is in the URL, but for amd64, its not there, so have that if else to substitute it anyways. i.e. for amd64, we can't use the arch_suffix.

tldr, even if we use the private bool function, we can update only one lie that is:
```
# For other arches, add it to the URL
        return f"{constants.RC_BASE_PRIV_URL.format(arch=arch)}/api/v1/releasestream/{tag}{arch_suffix}-priv"
```
to
```
# For other arches, add it to the URL
        return f"{constants.RC_BASE_PRIV_URL.format(arch=arch)}/api/v1/releasestream/{tag}{arch_suffix}"
```

File: doozer/doozerlib/cli/get_nightlies.py (hunk 4 above)
Comment (ashwindasr) [reply doozer/doozerlib/cli/get_nightlies.py:279]: sadly no - we would have to check for private nightly + amd64, and if true, we need to skip inserting the arch to the URL

File: doozer/doozerlib/cli/get_nightlies.py (hunk 1 above)
Comment (ashwindasr) [reply doozer/doozerlib/cli/get_nightlies.py]: done

File: doozer/doozerlib/cli/get_nightlies.py (hunk 3 above)
Comment (ashwindasr) [reply doozer/doozerlib/cli/get_nightlies.py]: done

File: doozer/doozerlib/cli/get_nightlies.py (hunk 4 above)
Comment (ashwindasr) [reply doozer/doozerlib/cli/get_nightlies.py]: done

File: doozer/doozerlib/constants.py (hunk 5 above)
Comment (ashwindasr) [reply doozer/doozerlib/constants.py]: done

File: doozer/doozerlib/cli/get_nightlies.py (hunk 6 above)
Comment (ashwindasr) [reply doozer/doozerlib/cli/get_nightlies.py]: done

File: doozer/doozerlib/constants.py (hunk 5 above)
Comment (ashwindasr) [reply doozer/doozerlib/constants.py]: Thanks for finding the URL, makes things much easier now

**Comment by ashwindasr** [comment]
retested with gen-assembly job, works as expected

File: doozer/doozerlib/cli/get_nightlies.py [hunk 7]
```diff
@@ -187,11 +193,26 @@ async def find_rc_nightlies(runtime: Runtime, arches: Set[str], allow_pending: b
...
-        rc_url: str = f"{rc_api_url(tag_base, _arch)}/tags"
+        rc_url: str = f"{rc_api_url(tag_base, _arch, private_nightly)}/tags"
         logger.info(f"Reading nightlies from {rc_url}")
 
+        headers = {}
+        if private_nightly:
+            # Get the token
+            rc, token, err = exectools.cmd_gather(["oc", "whoami", "-t"], strip=True)
+
+            if rc:
```
Comment (thegreyd) [inline doozer/doozerlib/cli/get_nightlies.py:204]:
```suggestion
            if rc != 0:
```

File: doozer/doozerlib/cli/get_nightlies.py [hunk 8]
```diff
@@ -187,11 +193,26 @@ async def find_rc_nightlies(runtime: Runtime, arches: Set[str], allow_pending: b
...
+                raise ValueError("Error while trying to get the token for reading private nightlies")
```
Comment (thegreyd) [inline doozer/doozerlib/cli/get_nightlies.py:205]:
```suggestion
            if rc != 0 or err:
                raise ValueError(f"Error while trying to get the token for reading private nightlies: {err}")
```

Comment (ashwindasr) [reply doozer/doozerlib/cli/get_nightlies.py]: thanks, updated
